### ISSUE TITLE
over aggressive line breaks

### ISSUE THREAD

**Issue by todd-richmond** [Camelcade/Perl5-IDEA#2166]
OSX with Intellij 2020.1 and Perl Plugin 2020.1

There are many cases where reformatting causes lines to be broken too early. My goal is to wrap as close to 80 chars as possible after an operator so that code is as dense as possible. However, reformatting will often break in an attempt to keep lists, ternaries or other items together on a new line even when that isn't helpful (or even if the next line is > 80). For example this simple case will be broken after the '=' instead of after $ccccccc,  It happens quite a bit more often but this is a simple fake one

Perltidy has an option to control this a bit better but I couldn't find anything similar in Perl5 config

@array = ($aaaaaaaaaaaaaaaa, $bbbbbbbbbbbbbbbb, $ccccccccccccccccccccccc, $ddddddddd);

**Comment by hurricup** [comment]
Seems this is default logic of IDEA formatter. I'll investigate if we can control this. 
For now I can only suggest to disable wrapping for assignment expression.

**Comment by todd-richmond** [comment]
Ah - I did forget to mention that I have 'wrap if long' set for all wrap options. I'll try disabling, though that might prevent breaks at 80

**Comment by hurricup** [comment]
If you disable all wrappings - no wrappings will be performed. 
The problem here is how to choose where to wrap. Plugin has 2 possibilities to wrap - assign expression or right part of assignment (comma sequence). And seems by default it chooses assign expression. 
It's not a simple text wrapper. It works with syntax tree:
```
assignment:
    left_part
    '='
    comma_sequence  # or wrap here
                 item1
                ','
                item2     # wrap here
                ','
....
```

**Comment by todd-richmond** [comment]
Just checked and I think this setting is probably good enough if I set every option to 'wrap if long' except assignment. That allows ternaries, operators and method calls to wrap as desired even on an assignment line
I also had to do 'chop down if long' for binary expressions in some cases to get shorter operator chains to wrap at the operator - otherwise the entire set would still wrap after the =
That fixed most, though not every instance I saw, but I think you can close this issue. It still works far better than perltidy after this change and other fixes unless you see a fixable issue with binary expressions

**Comment by hurricup** [comment]
I'll keep it until wrapping case is clear. Glad you found a suitable solution.

**Comment by hurricup** [comment]
Looks like this is kind of platform feature. Wrapping depends on how far element goes behind the wrap limit. And this looks strange. Here is the example for hard wrap 120:
```
@array = ($aaaaaaaaaaaaaaaa, $bbbbbbbbbbbbbbbb, $ccccccccccccccccccccccc, $ccccccccccccccccccccccc, $cccxcccxxcccccccccccccccc, $ddddddddd);

@array = ($aaaaaaaaaaaaaaaa, $bbbbbbbbbbbbbbbb, $ccccccccccccccccccccccc, $ccccccccccccccccccccccc, $cccxcccxxcccccccccccccc, $ddddddddd);
```

After re-formatting we get

```
@array = ($aaaaaaaaaaaaaaaa, $bbbbbbbbbbbbbbbb, $ccccccccccccccccccccccc, $ccccccccccccccccccccccc,
    $cccxcccxxcccccccccccccccc, $ddddddddd);
@array =
    ($aaaaaaaaaaaaaaaa, $bbbbbbbbbbbbbbbb, $ccccccccccccccccccccccc, $ccccccccccccccccccccccc, $cccxcccxxcccccccccccccc,
        $ddddddddd);
```
In the second case we wrapped assignment, instead of pre-last list element.

**Comment by todd-richmond** [comment]
yep - that's what I see. if I manually wrap where I want though, a second reformat often works correctly. That allows me to do fixups w/o always breaking them on subsequent re-writes

**Comment by todd-richmond** [comment]
I also see the related issue that displays in your example above. The third line is indented another 4 spaces (8 total) instead of lining up at 4 char indent for every subsequent wrap. This happens to if statements as well as assignment

**Comment by hurricup** [comment]
yes, because second line indented relative to statement, and the third one relative to parenthesized expression.

**Comment by todd-richmond** [comment]
is there a way to just have a single indent regardless of any other line attributes? I just want to always indent 4 chars after wrapping at 80 (after operator) until the end of line - never add more indents. Otherwise code just keeps moving right until lines are uselessly short

**Comment by hurricup** [comment]
Not sure. This is the main idea of IDEA formatting. Code semantics affects it. It's not just a text packer. 
It can move to the absurd levels only if you have some long, deep expressions with multiple newlines on different depths.

**Comment by hurricup** [comment]
To be clear: it's doable for sure. But to achieve this, i need to treat each statement as list of words/symbols, instead syntax tree. And this looks to me as a step back, not forward. We are formatting code, not just a text after all.

**Comment by todd-richmond** [comment]
could it still be parsed as a syntax tree so that the break occurs after operators / (... but just skip attempting to keep statement groups together or incrementing indent after the first as 0 extra indent instead of another continuation indent?
that would convert
```
a &&
  (b) &&
     (c) &&
        (d)

to
a &&
  (b) &&
  (c) &&
  (d)

**Comment by hurricup** [comment]
It is work as your described now. The problem is nested expressions.

E.g. 
```
a && 
    b && 
    c > 
      c1 > 
      c2 > 
      c3 &&
    d
```
